### ISSUE TITLE
invalid tests

### ISSUE THREAD

**Issue by michaelficarra** [jquery/esprima#1502]
Each of these tests was flagged as invalid during a run of the [Shift parser](http://shift-ast.org/parser.html) over esprima's tests. I've hand verified them.

1) Early error: Functions with non-simple parameter lists may not contain a "use strict" directive
  
  ```
  {
    "key": "declaration/function/empty-param",
    "case": "function a([], []) {'use strict';}\n"
  }
  ```
2) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-as-default",
    "case": "export {foo as default};\n"
  }
  ```
3) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-as-specifier",
    "case": "export {foo as bar};\n"
  }
  ```
4) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-as-specifiers",
    "case": "export {foo as default, bar};\n"
  }
  ```
5) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-specifier",
    "case": "export {foo};\n"
  }
  ```
6) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-specifiers-comma",
    "case": "export {foo, bar,};\n"
  }
  ```
7) Early error: Exported binding "foo" is not declared
  
  ```
  {
    "key": "ES6/export-declaration/export-named-specifiers",
    "case": "export {foo, bar};\n"
  }
  ```
8) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/object-initialiser/invalid-proto-identifier-shorthand",
    "case": "({ __proto__: null, __proto__ })\n",
    "failure": {
      "index": 29,
      "lineNumber": 1,
      "column": 30,
      "message": "Error: Line 1: Duplicate __proto__ fields are not allowed in object literals",
      "description": "Duplicate __proto__ fields are not allowed in object literals"
    }
  }
  ```
9) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/object-initialiser/invalid-proto-literal-shorthand",
    "case": "({ \"__proto__\": null, __proto__ })\n",
    "failure": {
      "index": 31,
      "lineNumber": 1,
      "column": 32,
      "message": "Error: Line 1: Duplicate __proto__ fields are not allowed in object literals",
      "description": "Duplicate __proto__ fields are not allowed in object literals"
    }
  }
  ```
10) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/object-initialiser/invalid-proto-shorthand-identifier",
    "case": "({ __proto__, __proto__: null })\n",
    "failure": {
      "index": 23,
      "lineNumber": 1,
      "column": 24,
      "message": "Error: Line 1: Duplicate __proto__ fields are not allowed in object literals",
      "description": "Duplicate __proto__ fields are not allowed in object literals"
    }
  }
  ```
11) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/object-initialiser/invalid-proto-shorthand-literal",
    "case": "({ __proto__, \"__proto__\": null })\n",
    "failure": {
      "index": 25,
      "lineNumber": 1,
      "column": 26,
      "message": "Error: Line 1: Duplicate __proto__ fields are not allowed in object literals",
      "description": "Duplicate __proto__ fields are not allowed in object literals"
    }
  }
  ```
12) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/object-initialiser/invalid-proto-shorthands",
    "case": "({ __proto__, __proto__ })\n",
    "failure": {
      "index": 23,
      "lineNumber": 1,
      "column": 24,
      "message": "Error: Line 1: Duplicate __proto__ fields are not allowed in object literals",
      "description": "Duplicate __proto__ fields are not allowed in object literals"
    }
  }
  ```
13) [1:24]: Unexpected "0"
  
  ```
  {
    "key": "ES6/template-literals/literal-escape-sequences",
    "source": "`\\u{000042}\\u0042\\x42\\u0\\A\\0`"
  }
  ```
14) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/yield/invalid-yield-strict-function-declaration",
    "case": "function yield(){ \"use strict\"; }\n",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Use of future reserved word in strict mode",
      "description": "Use of future reserved word in strict mode"
    }
  }
  ```
15) esprima expects this valid program to fail
  
  ```
  {
    "key": "ES6/yield/invalid-yield-strict-function-expression",
    "case": "(function yield(){ \"use strict\"; })\n",
    "failure": {
      "index": 10,
      "lineNumber": 1,
      "column": 11,
      "message": "Error: Line 1: Use of future reserved word in strict mode",
      "description": "Use of future reserved word in strict mode"
    }
  }
  ```
16) Early error: Arrow parameters must not contain yield expressions
  
  ```
  {
    "key": "ES6/yield/yield-generator-arrow-default",
    "case": "function *g() { (x = yield) => {} }\n"
  }
  ```
17) Early error: Generator parameters must not contain yield expressions
  
  ```
  {
    "key": "ES6/yield/yield-generator-default-parameter",
    "case": "function *g(x = yield){}\n"
  }
  ```
18) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0201",
    "case": "function eval() {'use strict'; }",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
19) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0202",
    "case": "function arguments() {'use strict'; }",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
20) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0205",
    "case": "(function eval() {'use strict'; })()",
    "failure": {
      "index": 10,
      "lineNumber": 1,
      "column": 11,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
21) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0206",
    "case": "(function arguments() {'use strict'; })()",
    "failure": {
      "index": 10,
      "lineNumber": 1,
      "column": 11,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
22) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0208",
    "case": "(function package() {'use strict'; })()",
    "failure": {
      "index": 10,
      "lineNumber": 1,
      "column": 11,
      "message": "Error: Line 1: Use of future reserved word in strict mode"
    }
  }
  ```
23) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0234",
    "case": "function static() { \"use strict\"; }",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Use of future reserved word in strict mode"
    }
  }
  ```
24) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0235",
    "case": "function eval(a) { \"use strict\"; }",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
25) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0236",
    "case": "function arguments(a) { \"use strict\"; }",
    "failure": {
      "index": 9,
      "lineNumber": 1,
      "column": 10,
      "message": "Error: Line 1: Function name may not be eval or arguments in strict mode"
    }
  }
  ```
26) esprima expects this valid program to fail
  
  ```
  {
    "key": "invalid-syntax/migrated_0259",
    "case": "function x(...{ a }){}",
    "failure": {
      "index": 14,
      "lineNumber": 1,
      "column": 15,
      "message": "Error: Line 1: Unexpected token {"
    }
  }
  ```
27) [1:17]: Invalid variable declaration in for-in statement
  
  ```
  {
    "key": "statement/iteration/migrated_0019",
    "case": "for (var x = 42 in list) process(x);"
  }
  ```
28) [1:20]: Invalid variable declaration in for-in statement
  
  ```
  {
    "key": "statement/iteration/migrated_0021",
    "case": "for (var x = y = z in q);"
  }
  ```
29) [1:31]: Invalid variable declaration in for-in statement
  
  ```
  {
    "key": "statement/iteration/migrated_0022",
    "case": "for (var a = b = c = (d in e) in z);"
  }
  ```
30) [1:45]: Invalid variable declaration in for-in statement
  
  ```
  {
    "key": "statement/iteration/migrated_0023",
    "case": "for (var i = function() { return 10 in [] } in list) process(x);"
  }
  ```

(edited by @ariya: turning the list into a numbered one, for ease of reference)

**Comment by bakkot** [comment]
Note that the tests which are invalid due to initializers in for-in statements will [probably be correct in the next version of ECMAScript](https://github.com/tc39/ecma262/pull/614) (unfortunately).

**Comment by michaelficarra** [comment]
@bakkot I hope not 😞. Until the next meeting, we should assume things will stay as they are.

**Comment by nenionoda** [comment]
@michaelficarra  `function eval() {'use strict'; }` is not valid indeed

**Comment by mathiasbynens** [comment]
> `[1:1]: Unexpected "ૹ"`
> 
> ```
> {
>   "key": "ES2016/identifier/gujarati_zha",
>   "case": "ૹ\n\n"
> }
> ```

This test case is fine. [`ૹ` is a valid identifier.](https://mothereff.in/js-variables#%E0%AB%B9) It’s an `ID_Start` symbol that was added in Unicode 8.

**Comment by nenionoda** [comment]
@mathiasbynens you are completely right; but i guess @michaelficarra meant that is should actually parse in esprima, but it won't. (I am on esprima2.7.2 and can confirm it won't parse '\u0AF9' there.)

**Comment by nenionoda** [comment]
`function static() { 'use strict'; }` is most probably invalid; static is a reserved word in strict mode
my 50 cents

**Comment by mathiasbynens** [comment]
@nenionoda I had misunderstood. Thanks for clarifying!

**Comment by nenionoda** [comment]
@mathiasbynens apologies for my wrong remark; I just found out '\u0AF9' can not be parsed with shift either :\

**Comment by bakkot** [comment]
@nenionoda, per spec, strict mode code does not include the BindingIdentifier of a FunctionDeclaration:

> Code is interpreted as [strict mode code](https://tc39.github.io/ecma262/#sec-strict-mode-code) in the following situations: [...]

> Function code is strict mode code if the associated FunctionDeclaration [...] is contained in strict mode code or if the code that produces the value of the function's [[ECMAScriptCode]] internal slot begins with a Directive Prologue that contains a Use Strict Directive.

> [Function code](https://tc39.github.io/ecma262/#sec-types-of-source-code) is source text that is parsed to supply the value of the [[ECMAScriptCode]] and [[FormalParameters]] internal slots (see 9.2) of an ECMAScript function object.

Note that this last does *not* include the declaration's name.

So all of the tests having to do with using future reserved words as function names are indeed invalid. Likewise for `eval` and `arguments`.

--

shift-parser-js's Unicode support is [a bit behind](https://github.com/shapesecurity/shift-parser-js/issues/263) (sorry!), but IIRC [the Java version](https://github.com/shapesecurity/shift-java)'s Unicode support is better.

**Comment by michaelficarra** [comment]
@mathiasbynens Sorry, I was going by the info at http://www.fileformat.info/info/unicode/char/0af9/index.htm. I'll remove that issue from the list. Is there a more reliable and equally convenient Unicode ID_Start/ID_Continue reference I should use?

**PS**. Your link lists that as an ES6 / Unicode 8.0.0 identifier, but Unicode 8.0.0 was not *required* until ES2016. If it's just supposed to mean that an implementation is allowed to allow it as an identifier, then ES5 allows it as well.

**Comment by mathiasbynens** [comment]
> Is there a more reliable and equally convenient Unicode `ID_Start`/`ID_Continue` reference I should use?

I tend to use CodePoints.net by @Boldewyn, e.g. <https://codepoints.net/U+0AF9>. (Search for IDC or IDS on that page.)

> Your link lists that as an ES6 / Unicode 8.0.0 identifier, but Unicode 8.0.0 was not required until ES2016.

That’s correct. But note that even ES5 allowed implementations to support more recent Unicode versions, and IMHO it makes sense for parsers used for tooling to apply Postel’s law.

> If it's just supposed to mean that an implementation is allowed to allow it as an identifier, then ES5 allows it as well.

True, since it’s a symbol in the `Other_Letter` category.

**Comment by nenionoda** [comment]
@bakkot I rather highly suspect you might be mistaken (I also highly suspect I'm wrong). https://tc39.github.io/ecma262/#sec-function-definitions-static-semantics-early-errors says:

>  FunctionDeclaration:functionBindingIdentifier(FormalParameters){FunctionBody}
   FunctionDeclaration:function(FormalParameters){FunctionBody}
   FunctionExpression:functionBindingIdentifier(FormalParameters){FunctionBody} 
   ...
    If the source code matching this production is strict mode code, it is a Syntax Error if BindingIdentifier is the IdentifierName eval or the IdentifierName arguments.

**Comment by michaelficarra** [comment]
@nenionoda I believe @bakkot is right. The section he was quoting defines what is strict mode code. In your quoted early error, the "source code matching that production" is strict mode code only if it is contained within strict mode code because the BindingIdentifier is not part of the *function code*; only "source text that is parsed to supply the value of the [[ECMAScriptCode]] and [[FormalParameters]] internal slots" is part of the *function code*.

**Comment by bakkot** [comment]
This part of the spec has misled implementors and even other spec authors before. I've just opened the closely related tc39/ecma262#632.

**Comment by nenionoda** [comment]
@michaelficarra I guess I've **not** been wrong, judging from this:
> Function code is strict mode code if ... the code that produces the value of the function's [[ECMAScriptCode]] internal slot begins with a Directive Prologue that contains a Use Strict Directive.

Found it mentioned in @bakkot's https://github.com/tc39/ecma262/issues/632 and leaves me little doubt as to what "the source text matching [the production]" is. Please impeach the spec authors if you find me wrong. The spec's writing style in this particular part in particular, is neither exact nor understandable to me. The other is the meaning of '\8' and '\9', which I found nowhere in the spec :frowning:

**Comment by bakkot** [comment]
@nenionoda, to understand that paragraph you also need the definition of "function code", which I quote in that issue and above; it includes the body and parameters of the function but does *not* include the BindingIdentifier (nor the full FunctionDeclaration production, in my reading).

`\8` and `\9` are not given meaning anywhere in the spec because they do not match the grammar at all. This can be inferred by inspecting the `EscapeSequence` [section](https://tc39.github.io/ecma262/#prod-EscapeSequence) and the `LegacyOctalEscapeSequence` [extension](https://tc39.github.io/ecma262/#sec-additional-syntax-string-literals) to which it refers.

**Comment by mathiasbynens** [comment]
Relevant: https://bugs.ecmascript.org/show_bug.cgi?id=3477

**Comment by nenionoda** [comment]
@bakkot thanks for the \9 and \8 thing; DecimalDigit there was what actually made me ransack the spec for any mention of \8 and \9, which i could not come up with. Now I can be sure I've not missed anything( but it would have been more convenient had it been expressly mentioned in the spec.) Thanks again.

As for the issue of FunctionDeclaration, even though I agree with you that BindingIdentifier is not under the exact same scrutiny when in strict mode as the rest of the FunctionDeclaration's components, I can not agree with you that the early error caused by theFunctionDeclaration's BindingIdentifier being 'eval' or 'arguments' has nothing to do with the function's body having a strict mode directive. I wish a spec author could be contacted on that matter, as all our debate looks to arise from our different interpretations of what the "source text matching the production" is. :\

**Comment by nenionoda** [comment]
I will be highly surprised if 
> FunctionDeclaration : function BindingIdentifier ( FormalParameters ) { FunctionBody }
If the source code matching this production is strict mode code, the Early Error rules for StrictFormalParameters : FormalParameters are applied.
means this is an error:
```ecmascript
function Strict() {
   'use strict';
   // should it be an error? the source it is contained in is strict
  function containsDupeAndIsNonStrictButWillCauseErrorNevertheless(a,a) {}
}
```

rather than this:
```ecmascript
function Strict(a,a) { 
  'use strict'; 

}
```ecmascript

However my first example is most probably wrong, because
>   The function code of
a particular ECMAScript function does not include any source text that is parsed as the function code
of a nested FunctionDeclaration, FunctionExpression, GeneratorDeclaration, GeneratorExpression,
MethodDefinition, ArrowFunction, ClassDeclaration, or ClassExpression.

But again, this means the rule below
> FunctionDeclaration : function BindingIdentifier ( FormalParameters ) { FunctionBody }
If the source code matching this production is strict mode code, the Early Error rules for StrictFormalParameters : FormalParameters are applied.

could only be meaningful in an other strict mode code than a strict FunctionBody, like, for example, a module. That is what has made me question the obvious so much. Apologies, and thanks in advance.

**Comment by bakkot** [comment]
@nenionoda, your first example is certainly an error because the nested function is strict code, even though it is not part of the function code of the outer function. This is because

> Function code is strict mode code if the associated FunctionDeclaration [...] is contained in strict mode code [...]

The second example is intended to be an error, I'm pretty sure, but the spec as written implies it is not, which is the issue I opened at tc39/ecma262#632.

**Comment by nenionoda** [comment]
@bakkot thanks 
for the record though, the first example is most probably **not** an error; not at least as far as I understand from the spec, especially considering that
> The function code of a particular ECMAScript function does not include any source text that is parsed as the function code (nenionoda: that is, the params and the body) of a nested FunctionDeclaration, FunctionExpression, GeneratorDeclaration, GeneratorExpression, MethodDefinition, ArrowFunction, ClassDeclaration, or ClassExpression.

The latter is certainly an error, according to the spec.

**Comment by bakkot** [comment]
@nenionoda, the first (`function outer(){'use strict'; function inner(a, a){} }`) is an error for the reason I gave in my previous comment. Functions whose declarations are within strict code are themselves strict. The body of the nested function is not *function code* of the outer function, but its "associated FunctionDeclaration ... is contained in strict mode code" (i.e. the body of the outer function). This property is unrelated to whether it is function code of the outer function.

The other case is, per Allen, intended to be an error. It's not as the spec is written, and I expect we'll amend the spec to correct this.

**Comment by nenionoda** [comment]
@bakkot thanks
could I ask why the spec has said the saying below?
> The function code of a particular ECMAScript function **does not** include any source text that is parsed as the function code of a nested FunctionDeclaration, FunctionExpression, GeneratorDeclaration, GeneratorExpression, MethodDefinition, ArrowFunction, ClassDeclaration, or ClassExpression.

To me it means a function declaration in the body of another, strict, function does not get affected by the strict context. 

I  do appreciate that our debate is mostly due to below, which is almost contradicted by the above:
> Function code is strict mode code if the associated FunctionDeclaration [...] is contained in strict mode code [...]

This makes me almost certain the spec is faulty in this regard.

**Comment by bakkot** [comment]
@nenionoda, I expect your difficulty arises from conflating two unrelated questions, namely, whether a nested FunctionDeclaration (or rather its parameters and body) is included in the *function code* of the outer function (it is not) and whether said FunctionDeclaration is contained in the body of the outer function (it is).

These are not in conflict because they have nothing to do with each other. "Function code" is a specific term being defined, not a limitation on which source text is contained within the body of a function.

**Comment by nenionoda** [comment]
Thanks!
This totally resolved my confusion!

**Comment by ariya** [comment]
Thanks for the detailed analysis @michaelficarra! I guess this'll keep us busy for a while 😼

**Comment by bakkot** [comment]
The various "Invalid variable declaration in for-in statement" tests are [now legal](https://github.com/tc39/ecma262/pull/614#event-746917384) in sloppy mode only.

**Comment by ariya** [comment]
@michaelficarra Do you mind numbering each of those items for ease of reference? I can also do it/edit it if you are fine with it.

**Comment by michaelficarra** [comment]
@ariya Feel free. Note that for-in initialisers are now valid again :anguished:.

**Comment by ariya** [comment]
* Case 1: solved with #1677
* Case 2-7: tracked as #1579
* Case 8-12: solved with #1485
* Case 13: tracked as #1499
* Case 14,15,18-25: tracked as #1707
* Case 16: tracked as #1645
* Case 17: tracked as #1706
* Case 26: solved with #1681
* Case 27-30: only error in sloppy mode, see #1309 for details

**Comment by ariya** [comment]
This issue will be closed since some of the cases have been fixed and the rest have the corresponding individual defect report.
